equinox/nn/_dropout.py

```diff
@@ -83,15 +88,24 @@ def __call__(
 
         if inference is None:
             inference = self.inference
-        if isinstance(self.p, (int, float)) and self.p == 0:
-            inference = True
+
         if inference:
             return x
-        elif key is None:
-            raise RuntimeError(
-                "Dropout requires a key when running in non-deterministic mode."
-            )
         else:
-            q = 1 - lax.stop_gradient(self.p)
+            p = self.p
+            if key is None:
+                p = error_if(
+                    p,
+                    p != 0,
+                    "Dropout requires a key when running in non-"
+                    "deterministic mode with non-zero probability.",
+                )
```

Comment:
Ack, I hadn't really thought about the fact that we're using using `error_if` here. This is actually pretty slow + sometimes buggy on TPUs etc. It's very much a tool of last resort: as such I try to reserve its usage for the scientific ecosystem stuff (Diffrax etc.) and up till now it has been entirely unused in `eqx.nn`.

I'm not sure what our possibilities are here -- can this be avoided? I'm tempted to say that the status quo is better than using `error_if`.